src/devices/bus/sg1000_exp/sp400.cpp

```diff
@@ -0,0 +1,210 @@
+{
+	m_dserdata = v;
+}
+
+void sp400_device::dser_p2_w(uint8_t v)
+{
+	m_busy = BIT(v, 5);
+	m_dserstrobe = BIT(v, 6);
+	m_motcpu->set_input_line(M6805_IRQ_LINE, m_dserstrobe);
+}
+
+uint8_t sp400_device::dser_p2_r()
+{
+	return
+		(m_motbusy<<7)|
+		(m_dserstrobe<<6)|
+		(m_busy<<5)|
+		(m_data<<4)|
+		(1<<3)|
+		(1<<2)|
+		(1<<1)|
+		(1<<0);
+}
+
+uint8_t sp400_device::mot_pa_r()
+{
+	return
+		(0<<0)|
+		(0<<1)|
+		(m_frontbuttons->read()<<2)| // bits 2,3 and 4
+		(0<<5)|
+		(0<<6);
+}
+
+void sp400_device::mot_pb_w(uint8_t v)
+{
+	m_plotter->update_motors(v&0x0F, (v>>4)&0x0F);
+}
+
+
+uint8_t sp400_device::mot_pc_r()
+{
+	return
+		(m_motPenUp<<0)|
+		(m_motPenDown<<1)|
+		(1<<2)|
+		(m_plotter->get_reedswitch_state()<<3)|
+		(m_motbusy<<4)|
+		(1<<5)|(1<<6)|(1<<7); //unconfirmed
+}
+
+void sp400_device::mot_pc_w(uint8_t v)
+{
+	m_motPenUp = BIT(v, 0);
+	m_motPenDown = BIT (v, 1);
+	m_motbusy = BIT(v, 4);
+	penZ();
+}
+
+uint8_t sp400_device::mot_pd_r()
+{
+	return sp400_device::m_dserdata;
+}
+
+void sp400_device::penZ()
+{
+	if (!m_motPenDown)
+	{
+		m_plotter->pen_down();
+	}
+	if (!m_motPenUp)
+	{
+		m_plotter->pen_up();
+	}
+}
+
+void sp400_device::device_add_mconfig(machine_config &config)
+{
+	i8035_device& dsercpu(I8035(config, m_dsercpu, 4_MHz_XTAL));
+	dsercpu.p1_out_cb().set(FUNC(sp400_device::dser_p1_w));
+	dsercpu.p2_in_cb().set(FUNC(sp400_device::dser_p2_r));
+	dsercpu.p2_out_cb().set(FUNC(sp400_device::dser_p2_w));
+	dsercpu.set_addrmap(AS_PROGRAM, &sp400_device::dser_map);
+	m6805_hmos_device& motcpu(M6805U3(config, m_motcpu, 4_MHz_XTAL));
+	motcpu.porta_r().set(FUNC(sp400_device::mot_pa_r));
+	motcpu.portb_w().set(FUNC(sp400_device::mot_pb_w));
+	motcpu.portc_r().set(FUNC(sp400_device::mot_pc_r));
+	motcpu.portc_w().set(FUNC(sp400_device::mot_pc_w));
+	motcpu.portd_r().set(FUNC(sp400_device::mot_pd_r));
+	generic_plotter_device& plotter(ALPS_DPG1302(config, m_plotter));
+	plotter.set_panel_update(FUNC(sp400_device::update_panel));
+}
+
+uint32_t sp400_device::update_panel(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+	bitmap.plot_box(100, 900-30, 90, 30, 0x00C000);
+	bitmap.plot_box(300, 900-30, 90, 30, m_motbusy?0xFF0000:0x100000);
+	int buttons=m_frontbuttons->read();
+	static constexpr uint32_t NORMALCOLOR=0x808080;
+	static constexpr uint32_t PRESSEDCOLOR=0x404040;
+	bitmap.plot_box(500,900-30, 90, 30, buttons&0x02?NORMALCOLOR:PRESSEDCOLOR);
+	bitmap.plot_box(700,900-30, 90, 30, buttons&0x01?NORMALCOLOR:PRESSEDCOLOR);
+	bitmap.plot_box(900,900-30, 90, 30, buttons&0x04?NORMALCOLOR:PRESSEDCOLOR);
+	return 0;
+}
+
+INPUT_CHANGED_MEMBER(sp400_device::misc_interaction)
+{
+	if (BIT(m_misc->read(), 0) == 0)
+	{
+		m_plotter->write_snapshot_to_file();
+	}
+	if (BIT(m_misc->read(), 1) == 0)
+	{
+		m_plotter->change_paper();
+	}
+}
+
+ROM_START( sp400 )
+ROM_REGION( 0x1000, "dsercpu", 0 )
+ROM_LOAD( "sp400_8035.bin", 0x0000, 0x1000, CRC(0eb48272) SHA1(08a2727f1592f5d2ecb2e368126ad7bfc5d3c270) )
```

Comment:
This is misaligned, we generally add 4 spaces/1 tab for each `ROM_REGION` / `ROM_LOAD` macros

Reply:
fixed